src/test/fuzz/sanitizestring.cpp

```diff
@@ -0,0 +1,90 @@
+
+// SEGV on SanitizeString()
+
+// If the attacker can control the 2nd arg of SanitizeString it segfaults (reasonably)
+// through call cross-reference, SanitizeString is  used on the codebase for printing/debugging
+ 
+// possible attack surface where the control of the 2nd arg of SanitizeString seems to be controllable:
+
+// [1]
+/* net.cpp:
+ *   void CConnman::PushMessage(CNode* pnode, CSerializedNetMsg&& msg) {
+ *       size_t nMessageSize = msg.data.size();    <---------- // [ No boundary check for nMessageSize passed to SanitizeString ]
+ *       LogPrint(BCLog::NET, "sending %s (%d bytes) peer=%d\n",  SanitizeString(msg.m_type), nMessageSize, pnode->GetId()); <----- 
+ */
+
+ // [2]
+ /* net.cpp
+  * Another method where the `message size` is passed to SanitizeString without check 
+  * std::optional<CNetMessage> V1TransportDeserializer::GetMessage(...)
+  *
+  */
+
+// NOTE: I do not know if the attacker can actually control &msg at this point, nor how to hit a breakpoint on this path. 
+//       just thought important reporting because did not see a boundary check for nMessageSize on PushMessage()
+
+// Find the harness and ASAn log `as PoC for [1]`
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <stddef.h>
+#include <stdint.h>
+#include <string.h>
+#include <string>
+
+static const std::string CHARS_ALPHA_NUM = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789";
+
+static const std::string SAFE_CHARS[] =
+{
+    CHARS_ALPHA_NUM + " .,;-_/:?@()", // SAFE_CHARS_DEFAULT
+    CHARS_ALPHA_NUM + " .,;-_?@", // SAFE_CHARS_UA_COMMENT
+    CHARS_ALPHA_NUM + ".-_", // SAFE_CHARS_FILENAME
+    CHARS_ALPHA_NUM + "!*'();:@&=+$,/?#[]-_.~%", // SAFE_CHARS_URI
+};
+
+std::string SanitizeString(const std::string& str, int rule)
+{
+    std::string strResult;
+    for (std::string::size_type i = 0; i < str.size(); i++)
+    {
+        if (SAFE_CHARS[rule].find(str[i]) != std::string::npos)
+            strResult.push_back(str[i]);
+    }
+    return strResult;
+}
+
+extern "C" int LLVMFuzzerTestOneInput(const uint8_t *data, size_t size)
+{
+    std::string fuzzdata(reinterpret_cast<const char*>(data), size);
+    fuzzdata.push_back('\0');
+
+    SanitizeString(fuzzdata, fuzzdata.size());
```

Comment:
the second arg is an enum, not a size. See https://github.com/bitcoin/bitcoin/pull/6647#discussion_r38855960

Reply:
Also, this method is already fuzzed in 

```
src/test/fuzz/string.cpp:    (void)SanitizeString(random_string_1, fuzzed_data_provider.ConsumeIntegralInRange<int>(0, 3));

Reply:
See https://github.com/bitcoin/bitcoin/blob/master/doc/fuzzing.md#quickstart-guide on how to compile. To run, you'd have to pass `FUZZ=string ./src/test/fuzz/fuzz`. Let me know if you have any questions.

Reply:
I wrote this harness and reported to security@bitcoin without realizing you guys already wrote one 
However the second arg is passed as a size i.e. on [1] and [2] and other methods. See where SanitizeString is called and you can see different cases where the second arg is not an enum and thats why I referenced that the fix must be on the callers.
This issue was already confirmed by laawnj and he also confirmed that SanitizeString()'s second arg must be called as an enum but it does not happen in some cases but such cases are not so reliable as attack vectors so is not a priority fix.

Reply:
I am not sure what `[1]` and `[2]` refer to, but if it was called anywhere with size, we could tell by existing tests crashing.

Reply:
The only places I could find where the second parameter is passed at all is URI handling and UA comment handling, both of which look fine.

Reply:
Commit 26c06f24e5dcc32a7599abb8d670d22993c82bc2 removed the last use of FILENAME, so I guess that constant can be removed now